File: solutions/javascript/rotational-cipher/1/rotational-cipher.js
/**
 * Rotates the characters in the provided text by the given shift value.
 *
 * @param {string} text - The input text to be rotated.
 * @param {number} shift - The number of positions to shift each character.
 * @returns {string} - The rotated text.
 */
export function rotate(text, shift) {
  return [...text]
    .map((char) => {
      const isUpperCase = char.match(/[A-Z]/);
      const isAlphabetic = char.match(/[a-z]/i);
      const baseCharCode = (isUpperCase ? 'A' : 'a').charCodeAt(0);

      return isAlphabetic
        ? String.fromCharCode(
            ((char.charCodeAt(0) - baseCharCode + shift) % 26) + baseCharCode
          )
        : char;
    })
    .join('');
}

File: solutions/python/rotational-cipher/2/rotational_cipher.py
"""
This module provides a function to perform Caesar cipher rotation on text.
"""

import string


def rotate(text, key):
    """
    Rotate the characters in the given text by the specified key using a Caesar cipher.

    Parameters:
    text (str): The string to be rotated.
    key (int): The number of positions each letter in the text is shifted.

    Returns:
    str: The rotated string with each letter shifted by the given key.
    """
    def shift_char(character, key_, alphabet):
        """
        Shifts a character by a given key within a specified alphabet.
        Parameters:
        character (str): The character to be shifted.
        key_ (int): The shift key indicating how many positions to move the
                    character in the alphabet.
        alphabet (str): The string of characters representing the alphabet.

        Returns:
        str: The shifted character if it is in the alphabet, otherwise returns
             the original character.
        """
        if character in alphabet:
            return alphabet[(alphabet.index(character) + key_) % len(alphabet)]
        return character

    rotated_text = []
    for char in text:
        if char.isupper():
            rotated_text.append(shift_char(char, key, string.ascii_uppercase))
        elif char.islower():
            rotated_text.append(shift_char(char, key, string.ascii_lowercase))
        else:
            rotated_text.append(char)

    return "".join(rotated_text)
